### PR TITLE
single callback video event

### PR DESCRIPTION
**Pull request by dinitri**

### CODE REVIEW

File: src/android/CameraPreviewFragment.java [hunk 1]
```diff
@@ -258,6 +277,72 @@ public void hasFlash(HasFlashCallback hasFlashCallback) {
...
+        FileOutputOptions outputOptions = new FileOutputOptions.Builder(videoFile).build();
+
+        final Handler handler = new Handler(Looper.getMainLooper());
+        handler.postDelayed(new Runnable() {
+            @Override
+            public void run() {
+                if (recording != null) {
+                    recording.stop();
+                    recording = null;
+                }
```
Comment (dinitri) [inline src/android/CameraPreviewFragment.java:306]: we can stopVideoCapture directly

File: src/android/CameraPreviewFragment.java [hunk 2]
```diff
@@ -258,6 +278,68 @@ public void hasFlash(HasFlashCallback hasFlashCallback) {
...
+            recording.stop();
+            recording = null;
+            return;
+        }
+        UUID uuid = UUID.randomUUID();
+
+        String filename = uuid.toString() + ".mp4";
+        if (ActivityCompat.checkSelfPermission(this.getContext(), Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(this.getActivity(), new String[]{Manifest.permission.RECORD_AUDIO}, 200);
+        }
```
Comment (dinitri) [inline src/android/CameraPreviewFragment.java:292]: we can move this check at start of the method

File: src/android/CameraPreviewFragment.java [hunk 3]
```diff
@@ -258,6 +278,68 @@ public void hasFlash(HasFlashCallback hasFlashCallback) {
...
+                    } else if (videoRecordEvent instanceof VideoRecordEvent.Finalize) {
+                        VideoRecordEvent.Finalize finalizeEvent = (VideoRecordEvent.Finalize) videoRecordEvent;
+                        handler.removeCallbacksAndMessages(null);
+                        if (finalizeEvent.hasError()) {
+                            // Handle the error
+                            int errorCode = finalizeEvent.getError();
+                            Throwable errorCause = finalizeEvent.getCause();
+                            videoCallback.onError(errorCode + " " + errorCause);
+                        } else {
+                            // Handle video saved
```
Comment (dinitri) [inline src/android/CameraPreviewFragment.java:323]:
```suggestion
```

File: src/android/CameraPreviewFragment.java [hunk 4]
```diff
@@ -258,6 +278,68 @@ public void hasFlash(HasFlashCallback hasFlashCallback) {
...
+                            videoCallback.onStop(false, Uri.fromFile(videoFile).toString());
+                            Uri savedUri = finalizeEvent.getOutputResults().getOutputUri();
+                        }
+                        recording = null;
+                    }
+                    // Other event types can be handled if needed
```
Comment (dinitri) [inline src/android/CameraPreviewFragment.java:329]:
```suggestion
```

File: src/android/SimpleCameraPreview.java [hunk 5]
```diff
@@ -92,6 +103,91 @@ public boolean execute(String action, JSONArray args, CallbackContext callbackCo
...
+        return true;
+    }
+
+    private boolean startVideoCapture(CallbackContext callbackContext) {
+        if (fragment == null) {
+            callbackContext.error("Camera is closed");
+            return true;
+        }
+
+        // Example of sending an update to the JavaScript callback
```
Comment (dinitri) [inline src/android/SimpleCameraPreview.java:120]:
```suggestion
```

File: src/android/SimpleCameraPreview.java [hunk 6]
```diff
@@ -92,6 +103,91 @@ public boolean execute(String action, JSONArray args, CallbackContext callbackCo
...
+    }
+
+
+    private boolean stopVideoCapture(CallbackContext callbackContext) {
+        if (fragment == null) {
+            callbackContext.error("Camera is closed");
+            return true;
+        }
+
+        // Example of sending an update to the JavaScript callback
```
Comment (dinitri) [inline src/android/SimpleCameraPreview.java:182]:
```suggestion
```

File: src/android/SimpleCameraPreview.java [hunk 7]
```diff
@@ -267,6 +363,10 @@ public void fetchLocation() {
         }
     }
 
+
+
+
+
```
Comment (dinitri) [inline src/android/SimpleCameraPreview.java:369]:
```suggestion
```

File: www/SimpleCameraPreview.js [hunk 8]
```diff
@@ -1,45 +1,81 @@
-var exec = require("cordova/exec");
-var PLUGIN_NAME = "SimpleCameraPreview";
-var SimpleCameraPreview = function () {};
+  var exec = require("cordova/exec");
```
Comment (dinitri) [inline www/SimpleCameraPreview.js:1]: there is an indentation on the whole file, this is why the diff is huge

File: src/android/SimpleCameraPreview.java [hunk 9]
```diff
@@ -92,6 +103,89 @@ public boolean execute(String action, JSONArray args, CallbackContext callbackCo
         }
     }
 
+    private boolean initVideoCallback(CallbackContext callbackContext) {
+        this.videoCallbackContext = callbackContext;
+        PluginResult result = new PluginResult(PluginResult.Status.OK, "video cb initialized");
```
Comment (parveshneedhoo) [inline src/android/SimpleCameraPreview.java:108]:
```suggestion
        PluginResult result = new PluginResult(PluginResult.Status.OK, "video callback initialized");
```

File: www/SimpleCameraPreview.js [hunk 10]
```diff
@@ -43,3 +78,4 @@ SimpleCameraPreview.deviceHasUltraWideCamera = function (onSuccess, onError) {
 };
 
 module.exports = SimpleCameraPreview;
+
```
Comment (parveshneedhoo) [inline www/SimpleCameraPreview.js:81]:
```suggestion
```

File: www/SimpleCameraPreview.js [hunk 11]
```diff
@@ -2,6 +2,41 @@ var exec = require("cordova/exec");
...
+    return;
+  }
+
+  if (!SimpleCameraPreview.videoInitialized) {
+    exec(
+        (info) => {
+          SimpleCameraPreview.videoInitialized = true;
+          this.videoCallback(info);
+        } ,
+          (err) => {
```
Comment (HashirRajah) [inline www/SimpleCameraPreview.js:21]:
```suggestion
        (err) => {
```

File: www/SimpleCameraPreview.js [hunk 12]
```diff
@@ -2,6 +2,41 @@ var exec = require("cordova/exec");
...
+            this.videoCallback(null, err);
+        },
+        "SimpleCameraPreview",
+        "initVideoCallback",
+        []
+    );
+  }
+  exec(onSuccess, onError, PLUGIN_NAME, "startVideoCapture");
```
Comment (HashirRajah) [inline www/SimpleCameraPreview.js:29]:
```suggestion
  exec(() => {}, onError, PLUGIN_NAME, "startVideoCapture");
```

File: www/SimpleCameraPreview.js [hunk 13]
```diff
@@ -2,6 +2,41 @@ var exec = require("cordova/exec");
...
+    console.error("Call setVideoCallback first");
+    onError("Call setVideoCallback first");
+    return;
+  }
+
+  if (!SimpleCameraPreview.videoInitialized) {
+    exec(
+        (info) => {
+          SimpleCameraPreview.videoInitialized = true;
+          this.videoCallback(info);
```
Comment (HashirRajah) [inline www/SimpleCameraPreview.js:19]:
```suggestion
          onSuccess(info);
```

**Review by HashirRajah — APPROVED** [review]

File: www/SimpleCameraPreview.js [hunk 14]
```diff
@@ -16,17 +16,17 @@ SimpleCameraPreview.startVideoCapture = function (onSuccess, onError) {
     exec(
         (info) => {
           SimpleCameraPreview.videoInitialized = true;
-          this.videoCallback(info);
+          onSuccess(info);
```
Comment (YushraJewon) [inline www/SimpleCameraPreview.js:19]: this.videoCallback({videocb: "video initialiazed"});

File: www/SimpleCameraPreview.js [hunk 15]
```diff
@@ -2,6 +2,41 @@ var exec = require("cordova/exec");
...
+    console.error("Call setVideoCallback first");
+    onError("Call setVideoCallback first");
+    return;
+  }
+
+  if (!SimpleCameraPreview.videoInitialized) {
+    exec(
+        (info) => {
+          SimpleCameraPreview.videoInitialized = true;
+          this.videoCallback({ tvideoCallback: true });
```
Comment (dinitri) [inline www/SimpleCameraPreview.js:19]: tvideoCallback. ?
c'est pas plutot un truc du genre videoCallbackInitialized

File: www/SimpleCameraPreview.js [hunk 16]
```diff
@@ -2,6 +2,47 @@ var exec = require("cordova/exec");
...
+            (info) => {
+              if (info.videoCallbackInitialized) {
+                SimpleCameraPreview.videoInitialized = true;
+                onSuccess();
+              }
+              this.videoCallback(info);
+            } ,
+            (err) => {
+              onError(err);
+            },
```
Comment (zfir) [inline www/SimpleCameraPreview.js:39]:
```suggestion
            onError
```

File: www/SimpleCameraPreview.js [hunk 17]
```diff
@@ -2,6 +2,47 @@ var exec = require("cordova/exec");
 var PLUGIN_NAME = "SimpleCameraPreview";
 var SimpleCameraPreview = function () {};
 
+SimpleCameraPreview.videoInitialized = false;
+SimpleCameraPreview.videoCallback = null;
```
Comment (zfir) [inline www/SimpleCameraPreview.js:6]:
```suggestion
```

I think `videoInitialized` is enough and storing `videoCallback` as global is not required

File: src/android/CameraPreviewFragment.java [hunk 18]
```diff
@@ -258,6 +278,67 @@ public void hasFlash(HasFlashCallback hasFlashCallback) {
...
+                            Uri savedUri = finalizeEvent.getOutputResults().getOutputUri();
+                        }
+                        recording = null;
+                    }
+                });
+
+    }
+
+    public void stopVideoCapture() {
+        if (recording != null) {
```
Comment (dinitri) [inline src/android/CameraPreviewFragment.java:334]: I think we have to put this,
```suggestion
        if (recording != null && recording == true) {
```

**Pull request merged**